### ISSUE TITLE
support for structural pruning methods

### ISSUE THREAD

**Issue by aaronkl** [whittle-org/whittle#175]
**Is your feature request related to a problem? Please describe.**

We should provide an interface for structural pruning methods, such as N
pruning based on weight magnitude or methods like Wanda, which are commonly used in the literature and serve as practical baselines.

**Describe the solution you'd like**

Ideally, we would introduce a new, streamlined interface for these pruning methods. Given a model as input, the interface would set specific weights within structural components to zero according to a chosen heuristic. For example, the interface might look something like this:

```
class Pruner(object):

    def __call__(self, model: GPT,  **kwargs) -> float:
        model.reset_super_network()
        total_parameters = compute_parameters(model)
        self.prune(model, **kwargs)
        return self.count_sparse_parameters(model) / total_parameters

    def prune(self, model: GPT, **kwargs):
        return

    @staticmethod
    def count_sparse_parameters(model: GPT):
        params = 0
        for n,p in model.named_parameters():
            params = params + torch.sum(p.data!=0)
        return float(params)

class NMPruner(Pruner)

    def prune(self, model, n:int, m:int, **kwargs):
        ...



model = GPT(....)
pruner = NMPruner()
sr = pruner(model, n=2, m=4)

```

In the example above, we prune the model in-place for memory efficiency. We could also potentially provide some functionality to create a new checkpoint of a prune model, but this would require some additional sorting with the current whittle interface.

**Comment by aaronkl** [comment]
@rheasukthanker wdyt?

**Comment by rheasukthanker** [comment]
> @rheasukthanker wdyt?

LGTM. We might also want to support a eval_perplexity and eval_downstream function for the pruning baselines. Could be a function in the parent class, or we can create a separate evaluator to support this? Having a general evaluator would perhaps make more sense as we can then use it for models pruned with NAS too. Perhaps warrants a separate PR?

**Comment by aaronkl** [comment]
I agree, it's probably better to have that separate since it's not pruning specific. For downstream evaluation, we could just use our current interface to LLM Eval Harness right?
